### PR TITLE
cleanup obis observer. handle exception in platforms messages

### PR DESCRIPTION
**Pull request by Pnom**

### CODE REVIEW

File: tests/analog/commands.test.ts [hunk 1]
```diff
@@ -24,7 +25,9 @@ const checkExample = ( constructor: any, {parameters, config, hex: {header, body
     expect(command.parameters).toStrictEqual(parameters);
     expect(command.getParameters()).toStrictEqual(parameters);
     expect(command.toHex()).toBe(commandHex);
-    expect(command.toJson()).toBe(JSON.stringify(command.getParameters()));
+    if ( ! (command instanceof DataSegmentBase) ) {
```
Comment (DarkPark) [inline tests/analog/commands.test.ts:28]:
```suggestion
    if ( !(command instanceof DataSegmentBase) ) {
```

File: src/utils/constants/bytesEncodeFormat.ts [hunk 2]
```diff
@@ -0,0 +1,4 @@
+export const HEX = 0;
+export const BASE64 = 1;
+
+export type TBytesEncodeFormat = 0 | 1;
```
Comment (DarkPark) [inline src/utils/constants/bytesEncodeFormat.ts:4]: it's not a constant

File: src/utils/constants/bytesEncodeFormat.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+export const HEX = 0;
```
Comment (DarkPark) [inline src/utils/constants/bytesEncodeFormat.ts:1]: we already have constans dir in src

File: src/utils/bytesEncoding.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+import getHexFromBytes from './getHexFromBytes.js';
+import {hexFormatOptions, IHexFormatOptions} from '../config.js';
+import {TBytesEncodeFormat, HEX} from './constants/bytesEncodeFormat.js';
+
+
+export const getEncodingFromBytes = (
+    bytes: Uint8Array,
+    encodingFormat: TBytesEncodeFormat = HEX,
+    hexFormat: IHexFormatOptions = hexFormatOptions
+) => (encodingFormat === HEX
```
Comment (DarkPark) [inline src/utils/bytesEncoding.ts:13]:
```suggestion
) => (
    encodingFormat === HEX
```

?

File: src/utils/bytesEncoding.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    ? getHexFromBytes(bytes, hexFormat)
```
Comment (DarkPark) [inline src/utils/bytesEncoding.ts:14]: this code implies just a single `isHex` flag

File: src/analog/commands/DataSegmentBase.ts [hunk 6]
```diff
@@ -50,4 +53,13 @@ export default class DataSegmentBase extends Command {
 
         return new this(buffer.getDataSegment());
     }
+
+    toJson ( encodeFormat: TBytesEncodeFormat = HEX, hexFormat: IHexFormatOptions = {} ) {
```
Comment (DarkPark) [inline src/analog/commands/DataSegmentBase.ts:57]: interface is ambiguous

it allows to choose `base64` but still configure hex options

File: src/analog/Command.ts [hunk 7]
```diff
@@ -145,7 +146,8 @@ abstract class Command {
      *
      * @returns JSON string contains current parameters
      */
-    toJson () {
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    toJson ( encodeFormat: TBytesEncodeFormat = HEX, hexFormat: IHexFormatOptions = {} ) {
```
Comment (DarkPark) [inline src/analog/Command.ts:150]: why do we need this?

File: src/utils/bytesEncoding.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import getBase64FromBytes from './getBase64FromBytes.js';
+import getBytesFromBase64 from './getBytesFromBase64.js';
+import getBytesFromHex from './getBytesFromHex.js';
+import getHexFromBytes from './getHexFromBytes.js';
+import {hexFormatOptions, IHexFormatOptions} from '../config.js';
+import {TBytesEncodeFormat, HEX} from './constants/bytesEncodeFormat.js';
+
+
+export const getEncodingFromBytes = (
```
Comment (DarkPark) [inline src/utils/bytesEncoding.ts:9]:
```suggestion
export const getStringFromBytes = (
```
?

File: src/mtx/CommandBinaryBuffer.ts [hunk 9]
```diff
@@ -351,6 +357,20 @@ class CommandBinaryBuffer extends BinaryBuffer {
...
 
+    getFrameHeader (): IFrameHeader {
+        return {
+            type: this.getUint8(),
+            destination: this.getUint16(),
+            source: this.getUint16()
+        };
+    }
+
+    setFrameHeader ( header: IFrameHeader ) {
```
Comment (DarkPark) [inline src/mtx/CommandBinaryBuffer.ts:368]:
```suggestion
    setFrameHeader ( frameHeader: IFrameHeader ) {
```

for better readability

File: src/mtx/CommandBinaryBuffer.ts [hunk 10]
```diff
@@ -351,6 +357,20 @@ class CommandBinaryBuffer extends BinaryBuffer {
         super(dataOrLength, false);
     }
 
+    getFrameHeader (): IFrameHeader {
```
Comment (DarkPark) [inline src/mtx/CommandBinaryBuffer.ts:360]: for future use?

File: tests/analog/commands.test.ts (hunk 1 above)
Comment (Pnom) [reply tests/analog/commands.test.ts]: fixed

File: src/utils/constants/bytesEncodeFormat.ts (hunk 2 above)
Comment (Pnom) [reply src/utils/constants/bytesEncodeFormat.ts]: fixed

File: src/utils/constants/bytesEncodeFormat.ts (hunk 3 above)
Comment (Pnom) [reply src/utils/constants/bytesEncodeFormat.ts]: fixed

File: src/utils/bytesEncoding.ts (hunk 4 above)
Comment (Pnom) [reply src/utils/bytesEncoding.ts]: fixed

File: src/utils/bytesEncoding.ts (hunk 5 above)
Comment (Pnom) [reply src/utils/bytesEncoding.ts]: fixed

File: src/utils/bytesEncoding.ts [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+    bytes: Uint8Array,
+    encodingFormat: TBytesEncodeFormat = HEX,
+    hexFormat: IHexFormatOptions = hexFormatOptions
+) => (encodingFormat === HEX
+    ? getHexFromBytes(bytes, hexFormat)
+    : getBase64FromBytes(bytes));
+
+export const getBytesFromEncoding = ( data: string, encodingFormat: TBytesEncodeFormat = HEX ) => (encodingFormat === HEX
```
[the remark this answers is not in the document]
Comment (Pnom) [reply src/utils/bytesEncoding.ts]: fixed

File: src/utils/bytesEncoding.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (Pnom) [reply src/utils/bytesEncoding.ts]: can leave without

File: src/utils/index.ts [hunk 12]
```diff
@@ -11,4 +11,5 @@ export {default as getHexFromBytes} from './getHexFromBytes.js';
 export {default as getHexFromNumber} from './getHexFromNumber.js';
 export {default as mergeUint8Arrays} from './mergeUint8Arrays.js';
 export {default as BinaryBuffer} from './BinaryBuffer.js';
+export * from './bytesEncoding.js';
```
[the remark this answers is not in the document]
Comment (Pnom) [reply src/utils/index.ts]: fixed

File: src/utils/bytesEncoding.ts (hunk 8 above)
Comment (Pnom) [reply src/utils/bytesEncoding.ts]: entire file deleted

File: src/mtx/CommandBinaryBuffer.ts (hunk 9 above)
Comment (Pnom) [reply src/mtx/CommandBinaryBuffer.ts]: fixed

File: src/mtx/CommandBinaryBuffer.ts (hunk 10 above)
Comment (Pnom) [reply src/mtx/CommandBinaryBuffer.ts:362]: image
see in src/mtx/message.ts

File: src/analog/commands/DataSegmentBase.ts (hunk 6 above)
Comment (Pnom) [reply src/analog/commands/DataSegmentBase.ts]: fixed

File: src/analog/Command.ts (hunk 7 above)
Comment (Pnom) [reply src/analog/Command.ts]: fixed

File: src/utils/bytesConversion.ts [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+
+interface IBase64FormatOptions {}
+
+
+export type TBytesFormatOptions = IHexFormatOptions | IBase64FormatOptions;
+
+export const getStringFromBytes = (
+    bytes: Uint8Array,
+    conversionFormat: number = HEX,
+    hexFormat: TBytesFormatOptions = {}
```
Comment (DarkPark) [inline src/utils/bytesConversion.ts:17]:
```suggestion
    formatOptions: TBytesFormatOptions = {}
```

Comment (Pnom) [reply src/utils/bytesConversion.ts]: fixed

File: src/utils/bytesConversion.ts [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+
+
+interface IBase64FormatOptions {}
+
+
+export type TBytesFormatOptions = IHexFormatOptions | IBase64FormatOptions;
+
+export const getStringFromBytes = (
+    bytes: Uint8Array,
+    bytesFormat: number = HEX,
```
Comment (DarkPark) [inline src/utils/bytesConversion.ts:16]:
```suggestion
    bytesConversionFormat: number = HEX,
```
?

File: src/analog/commands/DataSegmentBase.ts [hunk 15]
```diff
@@ -50,4 +52,13 @@ export default class DataSegmentBase extends Command {
 
         return new this(buffer.getDataSegment());
     }
+
+    toJson ( conversionFormatId: number = HEX, conversionFormatOptions: TBytesFormatOptions = {} ) {
```
Comment (DarkPark) [inline src/analog/commands/DataSegmentBase.ts:56]: maybe move this method to parent Command?

Comment (Pnom) [reply src/analog/commands/DataSegmentBase.ts:56]: The parent class Command already has.
`toJson () {`
`    return JSON.stringify(this.getParameters());`
`}`

File: src/utils/bytesConversion.ts (hunk 14 above)
Comment (Pnom) [reply src/utils/bytesConversion.ts]: fixed

File: src/utils/bytesConversion.ts [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+import getBytesFromBase64 from './getBytesFromBase64.js';
+import getBytesFromHex from './getBytesFromHex.js';
+import getHexFromBytes from './getHexFromBytes.js';
+import {IHexFormatOptions} from '../config.js';
+
+
+interface IBase64FormatOptions {}
+
+
+export type TBytesFormatOptions = IHexFormatOptions | IBase64FormatOptions;
```
[the remark this answers is not in the document]
Comment (Pnom) [reply src/utils/bytesConversion.ts]: fixed

**Review by DarkPark — APPROVED** [review]

**Review by DeadByElpy — APPROVED** [review]

**Pull request merged**